core/src/com/jediterm/terminal/model/LinesBuffer.java

```diff
@@ -250,31 +204,27 @@ public void addLines(@NotNull List<TerminalLine> lines) {
 
   public void moveBottomLinesTo(int count, @NotNull LinesBuffer buffer) {
     count = Math.min(count, getLineCount());
-    buffer.addLinesFirst(myLines.subList(getLineCount() - count, getLineCount()));
-
-    removeBottomLines(count);
+    List<TerminalLine> lines = LinesStorageKt.removeFromBottom(myLines, count);
+    buffer.addLinesFirst(lines);
   }
 
   private void addLinesFirst(@NotNull List<TerminalLine> lines) {
-    List<TerminalLine> list = new ArrayList<>(lines);
-    list.addAll(myLines);
-    myLines = new ArrayList<>(list);
+    LinesStorageKt.addAllToTop(myLines, lines);
   }
 
   private void removeBottomLines(int count) {
-    myLines = new ArrayList<>(myLines.subList(0, getLineCount() - count));
+    LinesStorageKt.removeFromBottom(myLines, count);
   }
 
-  public int removeBottomEmptyLines(int bottomMostLineInd, int maxCount) {
+  public int removeBottomEmptyLines(int maxCount) {
+    int countToRemove = Math.min(maxCount, getLineCount());
```

Comment:
Seems `countToRemove` is redundant. Can we just use `maxCount` instead?

Reply:
Yes, looks like we can.